core/src/main/java/com/linecorp/armeria/common/AbstractHttpRequestBuilder.java

```diff
@@ -416,9 +429,14 @@ protected final HttpRequest buildRequest() {
 
     private RequestHeaders requestHeaders() {
         requestHeadersBuilder.path(buildPath());
+        final MediaType contentType = contentType();
+        if (contentType != null) {
+            requestHeadersBuilder.contentType(contentType);
+        }
         if (cookies != null) {
             requestHeadersBuilder.set(COOKIE, Cookie.toCookieHeader(cookies));
         }
+        final HttpData content = content();
         if (content == null || content.isEmpty()) {
             requestHeadersBuilder.remove(CONTENT_LENGTH);
```

Comment:
A user can set the content length with a `Publisher` based on the pre-calculation. 
So, I think we shouldn't remove it when `Publisher` is not null.